### ISSUE TITLE
table and figure cross-reference in word_document2

### ISSUE THREAD

**Issue by blueskypie** [rstudio/bookdown#746]
Could anyone help with the following problem?

for figures, the following produces the correct numbering, but no cross-reference (ignore the `\` in `\\````, the backslash is added to display the code in this post)
```
Figure \@ref(fig:figMAFdistribution) is ...

\```{r figMAFdistribution, fig.cap="my cap"}
    ....
\```
```

for tables, I have to use flextable since kable does not display complex tables in word correctly.  The following just produces `Table ??`
```
Table \@ref(tab:tabTest) is ...

\```{r tabTest}
    dataDf=data.frame(matrix(1:9,nrow = 3))
    ft=flextable(dataDf)
    set_caption(ft, "my cap")
\```
```

**Comment by cderv** [comment]
I don't thing flextable 📦  is fully compatible with how bookdown label and caption work. 
See for example this [SO question](https://stackoverflow.com/questions/56676952/table-cross-references-in-bookdown-with-ms-word-output)

You may try new version of flextable because there seems to be [an evolution](https://github.com/davidgohel/flextable/blob/737b66deef470635579eff1525fa0699e77ebbb4/NEWS#L10
) in `set_caption`
> `set_caption` can now be used with R Markdown for Word document and caption

have you this version ? 


Also, you'll find some info on how this work for bookdown in the book https://bookdown.org/yihui/bookdown/tables.html
> If you decide to use other R packages to generate tables, you have to make sure the label for the table environment appears in the beginning of the table caption in the form (\#label) (again, label must have the prefix tab:).


For the first issue, please follow the issue template and provide a reproductible example. The issue template also explain how post rmarkdown code in the issue; 

Thank you.

**Comment by blueskypie** [comment]
Thanks for the help, @cderv ! here is my rmd file:
````
---
title: "Untitled"
author: "John"
date: "7/18/2019"
output: 
  bookdown::word_document2:
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F)
```

## R Markdown

Figure \@ref(fig:figTest) is ...

```{r figTest, fig.cap="my cap"}
    plot(1:10,1:10)
```

Table \@ref(tab:tabTest) provided the results.

```{r tabTest, include= T, message = T, warning = T, echo= F, eval = T }
library(flextable)
dataDf=data.frame(matrix(1:9,nrow = 3))
dataDf=cbind(VID=c("top2","top2","3rd"),dataDf)
ft=flextable(dataDf)
set_caption(ft, "my cap")
```

Table \@ref(tab:tabTest1) provided the results.

```{r tabTest1, include= T, message = T, warning = T, echo= F, eval = T }
dataDf=data.frame(matrix(1:9,nrow = 3))
dataDf=cbind(VID=c("top2","top2","3rd"),dataDf)
ft=flextable(dataDf)
set_caption(ft, "Table: (\\#tab:tabTest1) my cap")
```

## session info
```{r}
xfun::session_info('bookdown')
```
````
Here is the output
[tmp (2).docx](https://github.com/rstudio/bookdown/files/3408325/tmp.2.docx)

**Comment by cderv** [comment]
Thanks for your input. 

I already answered for the flextable part. I let you see on that part. 

For the plot question
> but no cross-reference

Sorry but I can't see where the issue is. 
It seems the text part correctly reference the figure number
![image](https://user-images.githubusercontent.com/6791940/61491807-4ba5b500-a9b0-11e9-84c1-34e2d3441488.png)

I may have misunderstood something though... 🤔

**Comment by blueskypie** [comment]
Sorry for not making it clear! By cross-reference, I mean the clickable reference with in-bedded URL. 
For example, if the output is html, the `1` in `Figure 1` clickable and takes me to the figure.

**Comment by cderv** [comment]
Oh I see ! I am not really sure that bookdown allows clickable reference... currently, from the source code it will only put the correct number. The word generation is made by pandoc and I don't know if there is a way to make a cross reference with pandoc for word. 

This seems like a word feature you won't have for R generated word...
But I may be wrong, as I've never seen it yet.

About flextable, it is currently not compatible with how bookdown expect the caption to be build. That is why you see `??`. An issue could be open in `flextable` repos to try make it compatible with bookdown. 
> the caption must be written in the <caption></caption> tag

**Comment by blueskypie** [comment]
Thanks for the answer, @cderv ! I really hope bookdown::word_document2 can work as smoothly as html_document2. Here in my organization, the reports need to be reviewed and edited by many non-technical people, and Word is the only option.

**Comment by cderv** [comment]
> Here in my organization, the reports need to be reviewed and edited by many non-technical people, and Word is the only option.

You could also be interested in other R related work around word
- [officer](https://davidgohel.github.io/officer/), another approach without Rmarkdown 
- [redoc](https://noamross.github.io/redoc/), still in dev
- [officedown](https://github.com/davidgohel/officedown), still in dev too and in pause currently.

Also I discovered a tool in useR!2019 that could be of interest for you : [StatTag](https://sites.northwestern.edu/stattag/) - it links Rmardown and Word for facilitate collaboration within both wold. ([slide from user prez](http://user2019.r-project.org/static/pres/t257951.pptx))

hope it could help improve your workflow in complement of bookdown.

**Comment by cderv** [comment]
I note here some of my findings (mostly for myself) after digging a bit through how pandoc works with word output and what exists with cross reference for pandoc output. 

* There is a pandoc filter executable [pandoc-crossref](https://lierdakil.github.io/pandoc-crossref/). It is not fully compatible with all the choices in Rmarkdown and bookdown syntax but it helps understand what is possible. However, if install it can easily be used with any Rmarkdown document, considering the syntax for crossref filter is respected. (see its doc) This gives idea for a doc (or new helper format) that would allow easy reference in pandoc document output (with a system dependency on pandoc-crossref)
* By default, pandoc will [create bookmarks in docx](https://github.com/lierdakil/pandoc/blob/master/src/Text/Pandoc/Writers/Docx.hs) for any `id` in `div`, `span`, `code block` and `Figure`. I think there is something that can be done in bookdown single document output format to better use this feature. I'll work on that when I have time.
* For referencing table, a div with id should somehow wrap the table. Also Pandoc Markdon has [a caption feature](https://pandoc.org/MANUAL.html#extension-table_captions).
* There is definitely a feature request to make in flextable for compatibility. Currently, you need to use `kable`, `kableExtra`, `pander` or maybe [`huxtable`](https://hughjonesd.github.io/huxtable/index.html)

**Comment by atusy** [comment]
For a temporary workaround, `captioner` package can be used for cross referencing instead of the `bookdown`'s feature.

**Comment by blueskypie** [comment]
@cderv thanks for the detailed info! I really appreciate!
@atusy thanks for the tip as well! actually I'm doing exactly as you suggested.

**Comment by atusy** [comment]
I found a hack, but is not working perfectly.
Based on the cross referencing of `kable`output,
what we have to do is to prepend caption with `Table: (\\#tab:iris)` or something like that.
It becomes `Table: Table 1:`.
If the prepending keyword is only `(\\#tab:iris)` (removing `Table: `), cross reference is not working.


## Result

![image](https://user-images.githubusercontent.com/30277794/62755519-4c150580-baaf-11e9-900f-f0ae3805fb44.png)


## Source

````
---
output: bookdown::word_document2
---

Table \@ref(tab:iris) is about iris.

```{r, echo = FALSE}
flextable::set_caption(
  flextable::flextable(iris[1, ]),
  "Table: (\\#tab:iris) flextable"
)
```
````

**Comment by cderv** [comment]
Note (for myself mainly) : I think there is something to do with pandoc lua filter here. I work on something when I have time. ⏲

**Comment by atusy** [comment]
This feature is implemented in davidgohel/flextable#205

**Comment by cderv** [comment]
Awesome @atusy ! Thanks a lot !
